### PR TITLE
don't try to start a new event loop in WebAgg when in an ipykernel

### LINKED ISSUE #24119: [ENH]: don't try to start a new event loop in WebAgg when in an ipykernel
### Problem

Our typical workflow involves running WebAgg backend in ipykernels, where there is already an event loop running. In such cases MPL is still functional, but it throws a number of errors/warnings/exceptions during plt.show(), plt.figure() calls, when it tries to start an event loop. (it will fail, but it will keep functioning).
Screen Shot 2022-10-07 at 10 49 47 PM
Screen Shot 2022-10-07 at 10 50 01 PM
Screen Shot 2022-10-07 at 10 50 23 PM

### Proposed solution

In two functions use `asyncio.get_running_loop()` to check if a loop is already running. Only try to start a new one when there is no running loop. A PR implementing this will be submitted for review.

### PR DESCRIPTION
**Pull request by zhizheng1**
## PR Summary

Use `asyncio.get_running_loop()` to check if a loop is already running. Only try to start a new one when there is no running loop.

Fixes #24119

## PR Checklist

**Tests and Styling**
- [N/A] Has pytest style unit tests (and `pytest` passes).
- [N/A] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [N/A] New features are documented, with examples if plot related.
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [N/A] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Comment by tacaswell** [comment]
I am a bit concerned about the asyncio vs tornado contrast.  I do not recall of the top of my head when tornado rebased itself on top of asyncio (it was in the last few years, I think tornado 5 (?)), do we need to put a minimum on the supported versions of Tornado?

Can you comment on why this works?  I'm surprised that we are able to automatically graft onto the existing tornado loop and run the communication over it....

**Comment by zhizheng1** [comment]
I did a search and it seems it's been working since a very LONG time ago - see this issue #4042 from 2015 which had the same observation as mine. I think what this PR does is just those mentioned in the end of that issue - suppress the errors/exceptions about loop already running, and not touch what's already working.

File: lib/matplotlib/backends/backend_webagg.py [hunk 1]
```diff
@@ -245,6 +245,14 @@ def random_ports(port, n):
 
     @classmethod
     def start(cls):
+        try:
+            import asyncio
+            asyncio.get_running_loop()
+        except:
+            pass
```
Comment (tacaswell) [inline lib/matplotlib/backends/backend_webagg.py:252]:
```suggestion
        import asyncio
        try:
            asyncio.get_running_loop()
        except RuntimeError:
            pass
```

File: lib/matplotlib/backends/backend_webagg.py [hunk 2]
```diff
@@ -286,8 +294,12 @@ def ipython_inline_display(figure):
...
 
     WebAggApplication.initialize()
-    if not webagg_server_thread.is_alive():
-        webagg_server_thread.start()
+    try:
+        import asyncio
+        asyncio.get_running_loop()
+    except:
+        if not webagg_server_thread.is_alive():
+            webagg_server_thread.start()
```
Comment (tacaswell) [inline lib/matplotlib/backends/backend_webagg.py:302]:
```suggestion
    import asyncio
    try:
        asyncio.get_running_loop()
    except RuntimeError:
        if not webagg_server_thread.is_alive():
            webagg_server_thread.start()
```

**Comment by tacaswell** [comment]
Tornado 5 is indeed when tornado started to use the asyscio event loop if found, there is a static method `IOLoop.current()` which is how the running loop is injected into all of the relevant places where it needs to be found. 

We already document tornado >=5 so I think there is no further work needed here (but in a notebook you should probably be using `ipympl`).

**Review by tacaswell — APPROVED** [review]

**Review by QuLogic — APPROVED** [review]

**Comment by QuLogic** [comment]
Thanks @zhizheng1! Congratulations on your first PR to Matplotlib :tada: We hope to hear from you again.

**Pull request merged**